### PR TITLE
Support localization resolution from route prefix and cookie

### PR DESCRIPTION
**Pull request by atistrcsn**
Hi,

This PR just a draft code and an attempt to respond this issue #1 , but it is working for us. Can you please review it? Maybe it meets your initial criteria.
There is no tests so far (we are in a hurry these days) and not as optimized as it should be, but we could work on it later :)

The extensive use of `@KtorExperimentalAPI` annotation that bothers us mostly but it should do it for now.

Cheers.
Attila

@aymanizz: this is my first github PR (as far as I remember) so be gentle with me

### CODE REVIEW

**Comment by aymanizz** [comment]
Hi!

Thanks a lot for the PR, I'll have a look.

Also, no need to worry about optimization right now, if it works then that's good enough for me.

**Comment by atistrcsn** [comment]
Anyway, we are trying to use Thymeleaf templateing, but it's just doesn't work. Have you any experience with Ktor + Thymeleaf?
Using your i18n feature alone not helping, obviously :) I guess the ktor Thymeleaf implementation is in an early stage.
I have keys in i18n.Message properties, but Thymeleaf does not translate is, just insert the key to html.

I am wondering can I just add your `call.t(R("foobar"))` to the context of the template. Have you managed this?

File: src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt [hunk 1]
```diff
@@ -27,16 +49,71 @@ class I18n private constructor(configuration: Configuration): MessageResolver by
      *
      * [messageResolver] instance of MessageResolver used in message resolution. Defaults to
      * [ResourceBundleMessageResolver].
-     *
      * @see MessageResolver
```
Comment (aymanizz) [inline src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:52]: This should be at the end of the comment.

**Comment by aymanizz** [comment]
Thanks :smile: 

I haven't really used Ktor outside of building APIs, but I'd guess it should be possible to pass `call` as you said, have you tried it? If that doesn't work a temporary solution might be to construct a map of all the localized messages and pass that, possibly by implementing your own `MessageResolver` and extending it with such functionality.

**Comment by aymanizz** [comment]
What I originally had in mind was to use [route interception](https://ktor.io/advanced/pipeline/route.html). But this is certainly a good first step so I'll accept it and we can improve it in a later release.

There are a few style nits, and documentation that needs updating but I'll ignore these for now. I am also waiting for some tests, tests don't need to cover everything just the common use cases.

BTW, this will only work if all routes start with the locale prefix, right?

File: src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt [hunk 2]
```diff
@@ -1,20 +1,42 @@
...
     val defaultLocale = configuration.defaultLocale ?: supportedLocales.first()
 
-    init { Locale.setDefault(defaultLocale) }
+    val useOfCookie = configuration.useOfCookie
+    val localeCookieName = configuration.cookieName
+
+    val useOfRedirection = configuration.useOfRedirection
+    val supportedPathPrefixes = "(${supportedLocales.joinToString("|", transform = { it.language })})".toRegex()
+
+    @KtorExperimentalAPI
```
[the remark this answers is not in the document]
Comment (atistrcsn) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:34]: You are right. I'll remove them.

File: src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt [hunk 3]
```diff
@@ -27,16 +49,71 @@ class I18n private constructor(configuration: Configuration): MessageResolver by
...
     }
 
+    @KtorExperimentalAPI
+    private suspend fun intercept(ctx: PipelineContext<Unit, ApplicationCall>) {
+        val call = ctx.context
+        val language = call.locale.language
+
+        // todo: maybe using regex would be faster and simpler than working w/ lists
+        val uri = call.request.origin.uri.trimStart('/').trimEnd('/').split('/')
+        if (!uri.first().matches(supportedPathPrefixes)
```
[the remark this answers is not in the document]
Comment (atistrcsn) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:104]: You might be overlooked the `uri.trimStart('/')`, which eliminates this exception. I tried with / and it's woking.

[the remark this answers is not in the document]
Comment (atistrcsn) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:104]: Sorry for the bad english 'linguistique' :)

[the remark this answers is not in the document]
Comment (aymanizz) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:104]: Yup, you're right.

File: src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt [hunk 4]
```diff
@@ -46,32 +123,72 @@ class I18n private constructor(configuration: Configuration): MessageResolver by
...
+        val acceptLocales = request.acceptLanguage()
+        val ranges = if (acceptLocales != null) Locale.LanguageRange.parse(acceptLocales) else listOf()
+        var locale = Locale.filter(ranges, i18n.supportedLocales).firstOrNull()
             ?: Locale.lookup(ranges, i18n.supportedLocales)
             ?: i18n.defaultLocale
-}
+
+        var localeCookie: String? = null
+        if (i18n.useOfCookie) {
+            // cookie always overrides the accept language header
```
[the remark this answers is not in the document]
Comment (atistrcsn) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:168]: Good catch, we should reverse the code with returning locale.

But I think, we should keep this locale resolving order: 

1. URI language prefix OR pathParam (like `?lang=en`)
2. Session resolver (if using one)
3. Cookie resolver
4. Accept-Language header
5. Fixed defaults

I borrow the idea from some other well known i18n lib, like this (I used to work with Node):
[Nodejs i18n](https://github.com/mashpie/i18n-node/blob/fb9aa221d74a22ba81ebdc3f7bbc859fac93ccb4/i18n.js#L664)

In Spring, this is the default ordering, as I can tell.

What do you think?

**Comment by atistrcsn** [comment]
> What I originally had in mind was to use [route interception](https://ktor.io/advanced/pipeline/route.html). But this is certainly a good first step so I'll accept it and we can improve it in a later release.
> 
> There are a few style nits, and documentation that needs updating but I'll ignore these for now. I am also waiting for some tests, tests don't need to cover everything just the common use cases.


Yepp, Route pipelining will be the closest to the optimal solution. I'm learning Ktor for 2 days now, so I hadn't puzzled together allthe features it has :) It's a fascinating framework tought, and I love the kotlin language too. Nice mixture :D

Finishing this project in a week and I'll create some tests to these changes.

> BTW, this will only work if all routes start with the locale prefix, right?
Nope, in configuration, you can list excluding prefixes, like `/static` directory in our app.

File: src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (aymanizz) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:159]: I didn't mean to change the resolution order, just reorder the code to reflect that order. The current resolution order looks good to me.

[the remark this answers is not in the document]
Comment (aymanizz) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:159]: Also, thanks for pointing this out :slightly_smiling_face: I didn't think of it earlier, I think following other libraries convention is good.

[the remark this answers is not in the document]
Comment (atistrcsn) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:159]: > I didn't mean to change the resolution order, just reorder the code to reflect that order. The current resolution order looks good to me.

Ok than :sweat_smile:

[the remark this answers is not in the document]
Comment (atistrcsn) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:159]: I am that kind of reverse / inverse thinking programmer... we all hate them :rofl:

[the remark this answers is not in the document]
Comment (aymanizz) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:159]: lol. Well, this https://gist.github.com/aymanizz/2dd455b45c0ece64225135bd4ccdac3f looks more straightforward. It follows the order of the locale resolution.

**Comment by aymanizz** [comment]
If testing/documentation will block this, they can just be done in separate PRs. I'll merge this once you have resolved the above and cleaned up the comments in code.

**Comment by aymanizz** [comment]
Hey @atistrcsn any progress on this PR? This is blocking #4 since I don't want to introduce merge conflicts for your PR, was wondering if you have the time to address my comments :slightly_smiling_face:. Since these are small changes I could add them myself if you can't, you need to [allow edits from maintainers](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) though.

File: src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (atistrcsn) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:159]: We have a really short deadline with the currently running project, sorry for ignoring this PR a little longer than I wish.
Thank you for the gist, I copied it into my commit.

_Btw: nice catch with that labeled lambda return, I almost forget that it's a thing in kotlin :) Time to update my rusty kotlin._

**Comment by atistrcsn** [comment]
Hi, 

Trying to catch up. Thanks for that gist. It was a time saver for me.

**Comment by atistrcsn** [comment]
Is there anything else that I can do for you to merge this PR?

**Comment by aymanizz** [comment]
Haha, thanks a lot btw! Let me review it real quick.

File: src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt [hunk 5]
```diff
@@ -28,15 +48,65 @@ class I18n private constructor(configuration: Configuration): MessageResolver by
...
+        fun exclude(predicate: (call: ApplicationCall) -> Boolean) {
+            excludePredicates.add(predicate)
+        }
+    }
+
+    private suspend fun intercept(ctx: PipelineContext<Unit, ApplicationCall>) {
+        val call = ctx.context
+        val language = call.locale.language
+
+        // todo: maybe using regex would be faster and simpler than working w/ lists
```
Comment (aymanizz) [inline src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:96]: this should be removed.

**Comment by aymanizz** [comment]
LGTM

File: src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (aymanizz) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt]: No problem! :smile:

File: src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt (hunk 5 above)
Comment (aymanizz) [reply src/main/kotlin/com/github/aymanizz/ktori18n/I18n.kt:96]: Eh, its a non-issue so I'll ignore it.

**Pull request merged**